### PR TITLE
[WIP] Add Twincat ADS adapter

### PR DESCRIPTION
**Pull request by phectori**
Work on issue #13.

Adds an ADS adapter, to log signals via [Beckhoff Twincat ADS](https://infosys.beckhoff.com/english.php?content=../content/1033/tcadscommon/html/tcadscommon_intro.htm). 

Todo:
- [x] Add support for other variable types, only DINT and LREAL work now.
- [x] **Still needs to be tested, I have no PLC here.**

### CODE REVIEW

File: adapters/ads_adapter.py [hunk 1]
```diff
@@ -1,22 +1,167 @@
...
 
+class AdsClient:
+    Entry = namedtuple(
+        "Entry", ["name", "typename", "comment", "datatype", "datatype_size"]
+    )
 
+    # Some data types have same number, maybe use type string field found in __parse_ads_entries...
+    DATATYPE_MAP = {
+        17: pyads.PLCTYPE_BOOL,  # ADST_UINT8
+        17: pyads.PLCTYPE_BYTE,  # ADST_UINT8
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: This key is duplicate, it might make sense to reverse the dictionary so that the numbers are the values and the `pyads` types are the keys. Then this mapping can be derived from that mapping.

File: adapters/ads_adapter.py [hunk 2]
```diff
@@ -1,22 +1,167 @@
...
+        entries = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOAD, 0, ctypes.c_ubyte * upload_info.nSymSize
+        )
+
+        parsed_entries = self.__parse_ads_entries(
+            bytes(entries), upload_info.nSymSize, upload_info.nSymbols
+        )
+        return parsed_entries
+
+    def __format_timestamp(timestamp: datetime):
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: This method could be a function or a staticmethod, since self is not used.

Btw, using a double-underscore has some implications with inheritance, I suggest to use a single-underscore to mark the function as "private".

File: adapters/ads_adapter.py [hunk 3]
```diff
@@ -1,22 +1,167 @@
...
+        upload_info = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOADINFO,
+            0,
+            pyads.structs.SAdsSymbolUploadInfo,
+        )
+        return upload_info
+
+    def __parse_ads_entries(self, entries_data: bytes, size: int, count: int):
+        parsed_entries = []
+        total_bytes = 0
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: I was still thinking about this name, cursor or index would be better here.

File: adapters/ads_adapter.py [hunk 4]
```diff
@@ -1,22 +1,167 @@
...
+            comment = entries_data[
+                comment_offset : comment_offset + comment_len
+            ].decode("ascii")
+
+            total_bytes += entry_length
+            parsed_entries.append(
+                self.Entry(name, typ, comment, entry_datatype, entry_datatype_size)
+            )
+
+        assert len(parsed_entries) == count
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: Another assert could be: `assert total_bytes == size`

File: adapters/ads_adapter.py [hunk 5]
```diff
@@ -1,22 +1,167 @@
...
+            0,
+            pyads.structs.SAdsSymbolUploadInfo,
+        )
+        return upload_info
+
+    def __parse_ads_entries(self, entries_data: bytes, size: int, count: int):
+        parsed_entries = []
+        total_bytes = 0
+        while total_bytes < size:
+            fmt = "<6I3H"
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: The deserialization of a single entry could be a function named:
```python
def unpack_entry(data: bytes, index: int) -> (entry_size, entry):
   ...
```

File: adapters/ads_adapter.py (hunk 1 above)
Comment (phectori) [reply adapters/ads_adapter.py]: I was thinking of solving this by not using the ``datatype`` field of an entry. Since we only have to do this once (at ``add_device_notification``), we can also just use the ``typename`` field.

File: adapters/ads_adapter.py [hunk 6]
```diff
@@ -1,22 +1,173 @@
...
+
+    def _get_upload_info(self):
+        upload_info = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOADINFO,
+            0,
+            pyads.structs.SAdsSymbolUploadInfo,
+        )
+        return upload_info
+
+    def _unpack_entry(data: bytes, index: int) -> (int, Entry):
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: Since this function is unrelated to this class, it might make sense to move it out of the class and make it a simple function.

File: adapters/ads_adapter.py [hunk 7]
```diff
@@ -1,22 +1,173 @@
...
+        comment = entries_data[comment_offset : comment_offset + comment_len].decode(
+            "ascii"
+        )
+
+        return (
+            entry_length,
+            self.Entry(name, typ, comment, entry_datatype, entry_datatype_size),
+        )
+
+    def _unpack_entries(self, entries_data: bytes, size: int, count: int):
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: Since this function is unrelated to this class, it might make sense to move it out of the class and make it a simple function.

File: adapters/ads_adapter.py [hunk 8]
```diff
@@ -1,22 +1,173 @@
...
+import datetime
+import re
+import argparse
+from collections import namedtuple
+from typing import Type
 
-# TODO: implement.
 
+class AdsClient:
+    Entry = namedtuple(
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: Maybe we could give this thing a more descriptive name, such as `AdsVariableDescriptionEntry` or something along those lines.

File: adapters/ads_adapter.py [hunk 9]
```diff
@@ -1,22 +1,173 @@
...
+        self.plc = pyads.Connection(ams_net_id, ams_net_port)
+        self.plc.open()
+
+    def __del__(self):
+        for handle in self.notification_handles:
+            self.plc.del_device_notification(*handle)
+
+        self.plc.close()
+
+    def log(self, name: str, plc_type):
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: I was a bit confused by this name, since I thought it would log a logmessage. Maybe rename this to `subscribe_to_change_notifications` or something like that?

**Comment by phectori** [comment]
I incorporated some of the changes you suggested in the last commit. 

Most important is to test tomorrow with a PLC, I have no idea anymore whether the code even works.

File: adapters/ads_adapter.py [hunk 10]
```diff
@@ -1,22 +1,176 @@
...
+        handles = self.plc.add_device_notification(
+            name, attr, self.callback(plc_type), pyads.ADSTRANS_SERVERCYCLE
+        )
+        self.notification_handles.append(handles)
+
+    def subscribe(self, pattern: str):
+        parsed_entries = self.get_ads_entries()
+
+        for entry in parsed_entries:
+            if re.match(pattern, entry.name) is not None:
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: Why not use: 
``` if re.match(pattern, entry.name):```
?

File: adapters/ads_adapter.py [hunk 11]
```diff
@@ -1,22 +1,176 @@
...
+        upload_info = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOADINFO,
+            0,
+            pyads.structs.SAdsSymbolUploadInfo,
+        )
+        return upload_info
+
+    @staticmethod
+    def format_timestamp(timestamp: datetime):
+        return time.mktime(timestamp.timetuple()) + timestamp.microsecond / 1e6
```
Comment (windelbouwman) [inline adapters/ads_adapter.py]: I added this logic to the `LognplotTcpClient` class.
